### ISSUE TITLE
Evaluate default procs in the context of the instance

### ISSUE THREAD

**Issue by mikekreeki** [orgsync/active_interaction#217]
Virtus allows to use a result of a method call as a default value, like this for example:

```ruby
class AgeCalculator
  include Virtus.model

  attribute :date_of_birth, Date, default: :parse_date

  attribute :year
  attribute :month
  attribute :day

  def execute
    # ...
  end

  private

  def parse_date
    Date.new(year, month, day)
  end
end
```

This is helpful, since this simple class can be initialized with both Date object or year/month/day triplet.

One could try to rewrite this class using ActiveIntraction the following way:

```ruby
class AgeCalculator < ActiveInteraction::Base

  date :date_of_birth, default: -> { parse_date }

  integer :year, :month, :day

  def execute
    # ...
  end

  private

  def parse_date
    Date.new(year, month, day)
  end
end
```

ActiveInteraction allows to use procs as default values, which is helpful for values like `Date.today`, but the example above blows up:

```
NameError: undefined local variable or method `parse_date' for AgeCalculator:Class
```

- Is it an intention that the proc is evaluated in the context of the class itself? Wouldn't it be more useful to lazily evaluate default procs in the context of an instance once attributes are assigned?

**Comment by tfausak** [comment]
I would like to add this feature, but it will require a significant refactoring of ActiveInteraction's internals. #290 had the right idea, but it didn't work for hash inputs. I think this feature should be part of v3.0.0. You could argue that it could be released with a minor version bump, but it is technically changing the API we provide.

**Comment by tfausak** [comment]
I still want to add this feature. But it's going to be hard to implement. If anyone wants to take a crack at it, #290 is a good start.

**Comment by tfausak** [comment]
This has the potential to interact strangely with [hash defaults](https://github.com/orgsync/active_interaction/blob/v2.1.5/README.md#hash). Since hash defaults have to either be `nil` or `{}`, you couldn't really set a hash's default to the result of a proc. To wit:

``` rb
hash :x, default: {} do
  integer :a, default: 0
end
# x defaults to { a: 0 }

# This will not work.
hash :y, default: -> { x } do
  integer :a
end

# This will work.
hash :z, default: {} do
  integer :a, -> { x[:a] }
end
```

That will quickly get annoying if you have a lot of fields in the hash. 

This isn't a problem per se, but it's yet another oddity when it comes to default values for hashes.

**Comment by tfausak** [comment]
Even though this is on the v3.0.0 milestone, I think it could be added to v2.2.0. We do not guarantee which binding will be used when evaluating defaults. In spite of that, it is always `nil`, which is worthless. The only way the changes in #332 could break anything is if you relied on the default value throwing an error. That's not a reasonable use case.

**Comment by tfausak** [comment]
Y'all should be happy to hear that this feature has been released as part of ActiveInteraction v2.2.0. Sorry it took so long :smile:
